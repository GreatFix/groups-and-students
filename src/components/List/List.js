import React, { useState, useCallback, useEffect } from 'react'
import classes from './List.module.css'
import Cell from '../Cell/Cell'
import Pagination from '../Pagination/Pagination'

const List = (props) => {
  const [page, setPage] = useState(1)
  const [pageContent, setPageContent] = useState([])

  useEffect(() => {
    const temp = []
    for (let i = props.pagination * (page - 1); i < props.pagination * page; i++)
      props.array[i] && temp.push(props.array[i])
    setPageContent(temp)
  }, [page, props.array, props.pagination])

  const handleClickPaginationButton = useCallback((number) => {
    setPage(number)
  }, [])

  let paginationSize
  if (props.array.length % props.pagination) {
    paginationSize = props.array.length / props.pagination + 1
  } else {
    paginationSize = props.array.length / props.pagination
  }
  return (
    <div className={`${classes.List} ${props.className}`}>
      <ul>
        {pageContent.map((item) => {
          return (
            <Cell
              key={Math.random()}
              subContent={item[props.subProperty]}
              id={item[props.idProperty]}
              onClickLabel={props.onClickLabel}
              onClickEdit={props.onClickEdit}
              onClickDelete={props.onClickDelete}
            >
              {item[props.mainProperty]}
            </Cell>
          )
        })}
      </ul>
      <Pagination
        className={classes.Pagination}
        size={paginationSize}
        onClick={handleClickPaginationButton}
      />
    </div>
  )
}

export default List
