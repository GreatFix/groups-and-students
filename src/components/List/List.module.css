.List {
  padding: 8px;
  display: grid;
  grid-template-rows: 20fr 1fr;
  grid-template-columns: 80%;
  justify-content: center;
}
.List ul {
  overflow-y: auto;
}
.Add {
  display: flex;
}
.Add button {
  flex-grow: 1;
}
.Alt {
  font-size: 1.2em;
  text-align: center;
  margin: 4px 2px 0 2px;
  padding: 5px;
}
