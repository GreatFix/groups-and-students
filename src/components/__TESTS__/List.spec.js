import { screen, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import List from '../List/List'

const TEST_ID = 1
const TEST_NAME = 'TEST_NAME'
const TEST_LASTNAME = 'TEST_LASTNAME'
const TEST_CLASS = 'TEST_CLASS'

describe('List', () => {
  const array = [
    { id: TEST_ID, name: TEST_NAME, lastname: TEST_LASTNAME },
    { id: 2, name: 'TestName2', lastname: 'TestLastname1' },
    { id: 3, name: 'TestName3', lastname: 'TestLastname1' },
  ]
  describe('renders with array and mainProperty', () => {
    const { container } = render(<List array={array} mainProperty="name" />)
    const list = container.querySelector('.List')
    const ul = list.firstChild

    test('renders component', () => {
      expect(list).toBeInTheDocument()
    })
    test('renders ul', () => {
      expect(ul).toBeDefined()
    })
    test('renders li', () => {
      expect(ul.querySelectorAll('li').length).toBe(array.length)
    })
    test('render mainProperty text', () => {
      const firstLi = ul.firstChild
      const [main] = firstLi.childNodes
      expect(main.textContent).toBe(TEST_NAME)
    })
    test('render properly', () => {
      expect(list).toMatchSnapshot()
    })
  })

  describe('renders with array, mainProperty and subProperty', () => {
    const { container } = render(<List array={array} mainProperty="name" subProperty="lastname" />)
    const list = container.querySelector('.List')
    const ul = list.firstChild

    test('renders li', () => {
      expect(ul.querySelectorAll('li').length).toBe(array.length)
    })
    test('render mainProperty text', () => {
      const firstLi = ul.firstChild
      const [main] = firstLi.childNodes
      expect(main.textContent).toBe(TEST_NAME)
    })
    test('render subProperty text', () => {
      const firstLi = ul.firstChild
      const [, sub] = firstLi.childNodes
      expect(sub.textContent).toBe(TEST_LASTNAME)
    })
    test('render properly', () => {
      expect(list).toMatchSnapshot()
    })
  })

  describe('renders with array, mainProperty and indexOn', () => {
    const { container } = render(<List array={array} mainProperty="name" indexOn />)
    const list = container.querySelector('.List')
    const ul = list.firstChild

    test('renders li', () => {
      expect(ul.querySelectorAll('li').length).toBe(array.length)
    })
    test('render mainProperty text', () => {
      const firstLi = ul.firstChild
      const [, main] = firstLi.childNodes
      expect(main.textContent).toBe(TEST_NAME)
    })
    test('render index', () => {
      const firstLi = ul.firstChild
      const [index] = firstLi.childNodes
      expect(index.textContent).toBe('1')
    })
    test('render properly', () => {
      expect(list).toMatchSnapshot()
    })
  })

  describe('renders with array, mainProperty and pagination', () => {
    const PAGINATION = 2
    const { container } = render(<List array={array} mainProperty="name" pagination={PAGINATION} />)
    const list = container.querySelector('.List')
    const ul = list.firstChild

    test('renders li in the number of pagination', () => {
      expect(ul.querySelectorAll('li').length).toBe(PAGINATION)
    })
    test('render mainProperty text', () => {
      const firstLi = ul.firstChild
      const [main] = firstLi.childNodes
      expect(main.textContent).toBe(TEST_NAME)
    })

    test('render properly', () => {
      expect(list).toMatchSnapshot()
    })
  })

  describe('Clicks', () => {
    test('onClickAdd', () => {
      const handleClickAdd = jest.fn()
      render(<List array={array} mainProperty="name" onClickAdd={handleClickAdd} />)
      userEvent.click(screen.getAllByAltText('add')[0])
      expect(handleClickAdd).toHaveBeenCalledTimes(1)
    })

    test('onClickDelete', () => {
      const handleClickDelete = jest.fn((id) => id)
      render(<List array={array} mainProperty="name" onClickDelete={handleClickDelete} />)
      userEvent.click(screen.getAllByAltText('delete')[0])
      expect(handleClickDelete).toHaveBeenCalledTimes(1)
      expect(handleClickDelete.mock.results[0].value).toBe(1)
    })

    test('onClickEdit', () => {
      const handleClickEdit = jest.fn((id) => id)
      render(<List array={array} mainProperty="name" onClickEdit={handleClickEdit} />)
      userEvent.click(screen.getAllByAltText('edit')[0])
      expect(handleClickEdit).toHaveBeenCalledTimes(1)
      expect(handleClickEdit.mock.results[0].value).toBe(1)
    })

    test('onClickLabel', () => {
      const handleClickLabel = jest.fn((id) => id)
      render(<List array={array} mainProperty="name" onClickLabel={handleClickLabel} />)
      userEvent.click(screen.getByText(TEST_NAME))
      expect(handleClickLabel).toHaveBeenCalledTimes(1)
      expect(handleClickLabel.mock.results[0].value).toBe(1)
    })
  })

  test('gets a class', () => {
    const { container } = render(<List array={array} mainProperty="name" className={TEST_CLASS} />)
    expect(container.querySelector('.List')).toHaveClass(TEST_CLASS)
  })
})
