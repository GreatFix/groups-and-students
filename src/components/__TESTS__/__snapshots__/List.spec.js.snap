// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders with array and mainProperty render properly 1`] = `
<div
  class="List "
>
  <ul
    style="overflow-y: hidden;"
  >
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TEST_NAME
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TestName2
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TestName3
      </p>
    </li>
  </ul>
  <div
    class="Pagination Pagination"
  >
    <button
      class="Button "
      style="background-color: rgb(0, 72, 186);"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`List renders with array, mainProperty and indexOn render properly 1`] = `
<div
  class="List "
>
  <ul
    style="overflow-y: hidden;"
  >
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="Sub Index"
      >
        1
      </p>
      <p
        class="SimpleCell SimpleCell "
      >
        TEST_NAME
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="Sub Index"
      >
        2
      </p>
      <p
        class="SimpleCell SimpleCell "
      >
        TestName2
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="Sub Index"
      >
        3
      </p>
      <p
        class="SimpleCell SimpleCell "
      >
        TestName3
      </p>
    </li>
  </ul>
  <div
    class="Pagination Pagination"
  >
    <button
      class="Button "
      style="background-color: rgb(0, 72, 186);"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`List renders with array, mainProperty and pagination render properly 1`] = `
<div
  class="List "
>
  <ul
    style="overflow-y: hidden;"
  >
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TEST_NAME
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TestName2
      </p>
    </li>
  </ul>
  <div
    class="Pagination Pagination"
  >
    <button
      class="Button "
      style="background-color: rgb(0, 72, 186);"
      type="button"
    >
      1
    </button>
    <button
      class="Button "
      style="background-color: rgb(0, 72, 186);"
      type="button"
    >
      2
    </button>
  </div>
</div>
`;

exports[`List renders with array, mainProperty and subProperty render properly 1`] = `
<div
  class="List "
>
  <ul
    style="overflow-y: hidden;"
  >
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TEST_NAME
      </p>
      <p
        class="Sub Sub"
      >
        TEST_LASTNAME
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TestName2
      </p>
      <p
        class="Sub Sub"
      >
        TestLastname1
      </p>
    </li>
    <li
      class="Cell "
      style="transform: translateX(0%);"
    >
      <p
        class="SimpleCell SimpleCell "
      >
        TestName3
      </p>
      <p
        class="Sub Sub"
      >
        TestLastname1
      </p>
    </li>
  </ul>
  <div
    class="Pagination Pagination"
  >
    <button
      class="Button "
      style="background-color: rgb(0, 72, 186);"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;
